### ISSUE TITLE
Do not inherit from Driver, abstract Driver from establishing connection.

### ISSUE THREAD

**Issue by pavelfeldman** [GoogleChrome/lighthouse#798]
It is better to parametrize Driver with the client-provided connection that would implement following interface:

```
Connection {
  sendCommand(command, params): Promise
  on(eventName, cb) for eventName == 'notification' and 'closed'
}
```

it also would make sense to pass established connection into the gatherer and runner in order to abstract them from the connection specifics. Neither runner not gatherer should be able to either disconnect or re-established connection.

WDYT?

**Comment by brendankenny** [comment]
I think this would be great. The structure is like 80% there already, so I think this would make perfect sense.

My only questions (which may be cleared up best just by doing it :):
- there is some state associated with each type of driver. Some of it can be normalized, but some of it (like tab handling for the extension driver) is unique to each. Where would that live?
- Where are you seeing connection lifecycle management living? Outside the runner and passed in as `Driver` is currently? (done [here for node](https://github.com/GoogleChrome/lighthouse/blob/9c78853f0c6e86eae7383cc7f5ec9e3e6901dfa3/lighthouse-core/index.js#L64) and [here for the extension](https://github.com/GoogleChrome/lighthouse/blob/9c78853f0c6e86eae7383cc7f5ec9e3e6901dfa3/lighthouse-extension/app/src/lighthouse-background.js#L66))
- The unit tests of and using `Driver` are something of a landmine. The suggested model would actually help a ton with cleanly refactoring the tests, but the current tests may bite you with what they assume and how haphazardly some things are currently mocked.

**Comment by brendankenny** [comment]
fixed by #800
